src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -896,12 +896,24 @@ private static Vector128<ulong> Accumulate128(Vector128<ulong> accVec, byte* sou
             Vector128<uint> sourceKey = sourceVec ^ secret;
 
             // TODO: Figure out how to unwind this shuffle and just use Vector128.Multiply
-            Vector128<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 0, 3, 0));
             Vector128<uint> sourceSwap = Vector128.Shuffle(sourceVec, Vector128.Create(2u, 3, 0, 1));
             Vector128<ulong> sum = accVec + sourceSwap.AsUInt64();
-            Vector128<ulong> product = Sse2.IsSupported ?
-                Sse2.Multiply(sourceKey, sourceKeyLow) :
-                (sourceKey & Vector128.Create(~0u, 0u, ~0u, 0u)).AsUInt64() * (sourceKeyLow & Vector128.Create(~0u, 0u, ~0u, 0u)).AsUInt64();
+
+            Vector128<ulong> product;
+            // TODO: Workaround for ARM as `*` is software emulated.
+            if (System.Runtime.Intrinsics.Arm.AdvSimd.IsSupported)
+            {
+                Vector64<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(0u, 2, 0, 0)).GetLower();
+                Vector64<uint> sourceKeyHigh = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 3, 0, 0)).GetLower();
+                product = System.Runtime.Intrinsics.Arm.AdvSimd.MultiplyWideningLower(sourceKeyLow, sourceKeyHigh);
+            }
+            else
+            {
+                Vector128<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 0, 3, 0));
+                product = Sse2.IsSupported
+                    ? Sse2.Multiply(sourceKey, sourceKeyLow)
```

Comment:
I'm not sure I 100% understand what is going on here, but can we share `sourceKeyLow` and `sourceKeyHigh` for both platforms? I'm struggle to understand why shuffle masks are different. Here is the `sse2neon` impl for Arm's `Sse2.Multiply(V128<uint>, V128<uint>)`: https://github.com/DLTcollab/sse2neon/blob/159be6a2d08663a8063d259230d1cb3572fd6026/sse2neon.h#L4874-L4885

Reply:
Do you mean having a kind of `GetLower()` that picks up `(0, 2)` and `(1, 3)` while sharing the same input shuffle for both? `(0,-,2,-) x (1,-,3,-)`? I haven't looked if there is anything like that, but I would prefer to avoid a double shuffle just to make the "surface API" look the same. Or I'm missing something in your suggestion?

Reply:
No, what I meant is basically this:
```csharp
// Existing impl for Sse2
static Vector128<ulong> Product_Sse2(Vector128<uint> sourceKey)
{
    Vector128<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 0, 3, 0));
    return Sse2.Multiply(sourceKey, sourceKeyLow);
}

// Your ARM path
static Vector128<ulong> Product_Neon_Alexandre(Vector128<uint> sourceKey)
{
    Vector64<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(0u, 2, 0, 0)).GetLower();
    Vector64<uint> sourceKeyHigh = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 3, 0, 0)).GetLower();
    return AdvSimd.MultiplyWideningLower(sourceKeyLow, sourceKeyHigh);
}

// My suggestion (shares parts with x86 path) based on sse2neon
static Vector128<ulong> Product_Neon_Egor(Vector128<uint> sourceKey)
{
    Vector128<uint> sourceKeyLow = Vector128.Shuffle(sourceKey, Vector128.Create(1u, 0, 3, 0));
    return AdvSimd.MultiplyWideningLower(
        AdvSimd.ExtractNarrowingLower(sourceKeyLow.AsUInt64()),
        AdvSimd.ExtractNarrowingLower(sourceKey.AsUInt64()));
}
```
Not sure which one is faster though (and are they 100% semantically the same):
```asm
; Assembly listing for method C:Product_Neon_Alexandre
G_M51276_IG01:              ;; offset=0000H
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
G_M51276_IG02:              ;; offset=0008H
        9C000150          ldr     q16, [@RWD00]
        4E100010          tbl     v16.16b, {v0.16b}, v16.16b
        9C000191          ldr     q17, [@RWD16]
        4E110000          tbl     v0.16b, {v0.16b}, v17.16b
        2EA0C200          umull   v0.2d, v16.2s, v0.2s
G_M51276_IG03:              ;; offset=001CH
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
RWD00  	dq	0B0A090803020100h, 0302010003020100h
RWD16  	dq	0F0E0D0C07060504h, 0302010003020100h
; Total bytes of code 36


; Assembly listing for method C:Product_Neon_Egor
G_M54047_IG01:              ;; offset=0000H
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
G_M54047_IG02:              ;; offset=0008H
        4EA01C10          mov     v16.16b, v0.16b
        9C000131          ldr     q17, [@RWD00]
        4E110210          tbl     v16.16b, {v16.16b}, v17.16b
        0EA12A10          xtn     v16.2s, v16.2d
        0EA12800          xtn     v0.2s, v0.2d
        2EA0C200          umull   v0.2d, v16.2s, v0.2s
G_M54047_IG03:              ;; offset=0020H
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
RWD00  	dq	0302010007060504h, 030201000F0E0D0Ch
; Total bytes of code 40
```

Reply:
I double checked and surprisingly, your version is actually 25% slower, so I kept the original version.